art_internal.hpp

```diff
@@ -31,19 +31,51 @@ class [[nodiscard]] basic_db_leaf_deleter;
 // Internal ART key in binary-comparable format
 template <typename KeyType>
 struct [[nodiscard]] basic_art_key final {
+
+  // Convert an external key into an internal key supporting
+  // lexicographic comparison.  This is only intended for key types
+  // for which simple conversions are possible.  For complex keys,
+  // including multiple key components or Unicode data, the
+  // application should use a gsl::space<std::byte> which already
+  // supports lexicographic comparison.
   [[nodiscard, gnu::const]] static UNODB_DETAIL_CONSTEXPR_NOT_MSVC KeyType
   make_binary_comparable(KeyType key) noexcept;
 
+  // Convert an internal key into an external key. This is only
+  // intended for key types for which simple conversions are possible.
+  // For complex keys, including multiple key components or Unicode
+  // data, the application should use a gsl::space<std::byte> which
+  // already supports lexicographic comparison.
+  [[nodiscard, gnu::const]] static UNODB_DETAIL_CONSTEXPR_NOT_MSVC KeyType
+  make_external(KeyType key) noexcept;
+
   constexpr basic_art_key() noexcept = default;
 
   UNODB_DETAIL_CONSTEXPR_NOT_MSVC explicit basic_art_key(KeyType key_) noexcept
       : key{make_binary_comparable(key_)} {}
 
+  // @return true iff the two keys are equal.
   [[nodiscard, gnu::pure]] constexpr bool operator==(
       basic_art_key<KeyType> key2) const noexcept {
+    // FIXME This is wrong for variable length keys.  It needs to
+    // consider no more bytes than the shorter key and if the two keys
+    // have the same prefix, then they are != if one is longer (and
+    // for == we can just compare the size as a short cut for this).
+    // Also needs unit tests for variable length keys.
     return !std::memcmp(&key, &key2.key, size);
   }
 
+  // @return -1, 0, or 1 if this key is LT, EQ, or GT the other key.
+  [[nodiscard, gnu::pure]] constexpr int cmp(
+      basic_art_key<KeyType> key2 ) const noexcept {
+    const size_t n = size <= key2.size ? size : key2.size;
```

Comment:
Remove and just use `size` in `memcmp`. No reason to half-support varlength keys before they are a thing

Reply:
done